### ISSUE TITLE
First Ctrl+Scroll doesn't work

### ISSUE THREAD

**Issue by GabrielSampol** [elmarquis/Leaflet.GestureHandling#5]
First time that Ctrl+Mouse-wheel-scroll is used, the event seems to be caught by the Internet browser and the result is a change in the zoom level of the whole browser window, not in the zoom level of the map. The following scrolls are changing the map zoom level, as expected.

I am using Windows 10 and it happens both in Firefox and Chrome.

A simple test page I have made is at http://www.dxmaps.com/prueba.html

Am I doing something wrong or is there some workaround?

**Comment by elmarquis** [comment]
I've not seen that behaviour. Does the same thing happen for you on the Google version? https://developers.google.com/maps/documentation/javascript/examples/interaction-cooperative

**Comment by GabrielSampol** [comment]
No, with Google Maps JS API and enabling the cooperative gesture it works perfect.

 

Gabriel

 

De: elmarquis [mailto:notifications@github.com] 
Enviado el: jueves, 17 de mayo de 2018 17:01
Para: elmarquis/Leaflet.GestureHandling
CC: GabrielSampol; Author
Asunto: Re: [elmarquis/Leaflet.GestureHandling] First Ctrl+Scroll doesn't work (#5)

 

I've not seen that behaviour. Does the same thing happen for you on the Google version? https://developers.google.com/maps/documentation/javascript/examples/interaction-cooperative

**Comment by Turtlehead91** [comment]
I'm seeing the same thing in Chrome, however it isn't being 'caught' by the browser per-say.  It looks like the GestureingHand script is catching the event and THAN both the GesturingHand AND the browser are registering the scroll event.

**Comment by GabrielSampol** [comment]
I have made another test with Ubuntu and Firefox and the behaviour is the same, so the issue is not related to Windows.

**Comment by GabrielSampol** [comment]
As additional information:  The dragging of the map is also affected. When using GestureHandling I have to click on the map before the cursor becomes a hand and I can drag the map. Without GestureHandling it works as expected and clicking before dragging is not required.

Example with GestureHandling showing the issue at http://www.dxmaps.com/prueba.html
Example without GestureHandling at http://www.dxmaps.com/pruebax.html

**Comment by elmarquis** [comment]
Hi @GabrielSampol . Thanks for supplying great info. I've just pushed an update with e.preventDefault. 
I've been able to recreate the issue with the browser window zooming and found your suggested fix resolved it. 

Will look at some of the other things you've reported.

**Comment by elmarquis** [comment]
The issue with the map needing a click before the hand cursor appears is caused by this line: 

`this._map.dragging.disable();`

When the map is initialised, gesture handling deactivates scrolling and dragging so that they can only be activated for the duration of a valid ctrl+scroll event on desktop or two fingered pan on mobile. 

If I remove that line, the desktop experience works great and the hand cursor appears fine without needing a click. However, mobile or touch devices end up scrolling the map with a one fingered drag, while still showing the warning message. 

I'll need to have a think about this one. I could check for touch functionality and only disable dragging for those devices. But there's a number of windows laptops around which are touch enabled, yet most users just use the trackpad or mouse.

**Comment by Turtlehead91** [comment]
What about using a mouseOver event in parallel with the CTR to toggle the draggable? Touch devices only trigger a mouseOver event when the user holds down on a area. If a user holds down on the map, being able to drag it seems reasonable.

**Comment by elmarquis** [comment]
@Turtlehead91 That's a pretty good shout. I'll need to do a bit of cross device testing on it, but sounds like it could be the way to go. I'm done for the night just now, so will pick this up in a couple of days. In the meantime if anyone wants to experiment with this idea, feel free.

**Comment by elmarquis** [comment]
Making the hand cursor appear on desktop without needing to first click the map is a little trickier than I first thought. 
I tried the approach of using mouseover to enable dragging. 
It worked pretty good for desktop. 
As expected, on ios safari a longer hold will trigger mouseover which also enables dragging. However, even though the subsequent touch events disable dragging (I console logged to prove this was happening), the mouseover one seems to take precedence and the map remains scrollable with one finger until you find something else on the page to click on.
I also tried adding a click event on mouseover, but same deal. 

I'll have another think. Unfortunately there's no easy way to tell if a mouseover event came from an actual mouse or a long touch. 

Sniffing for touch functionality would be ropey too as plenty windows laptops have both touchscreen and mouse/keyboard.

**Comment by elmarquis** [comment]
Got a fix for this in branch v1.1.2. 

The baffling behaviour I was seeing was due to an undetected js error happening on mobile touch events. 

I still need to test on Android, once done I'll merge into master. 

If you get a chance to also test branch v1.1.2 before I merge it all helps.

**Comment by GabrielSampol** [comment]
Tried branch v1.1.2 with Firefox and Chrome, and the issue seems to have been fixed. Thanks!

**Comment by elmarquis** [comment]
Great thanks. Now merged into master. Closing this issue.
